File: app/src/main/java/ro/pub/cs/systems/eim/practicaltest02v2/view/PracticalTest02v2ServerActivity.java
package ro.pub.cs.systems.eim.practicaltest02v2.view;

import android.os.Bundle;
import android.widget.TextView;
import android.support.v7.app.AppCompatActivity;

import java.io.InputStream;
import java.net.Socket;

import ro.pub.cs.systems.eim.lab07.calculatorwebservice.R;

public class PracticalTest02v2ServerActivity extends AppCompatActivity {

    private boolean isRunning = true; // Flag to control the connection loop

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_practical_test02v2_server);

        TextView textView = findViewById(R.id.textView);

        new Thread(() -> {
            try {
                // Replace with your laptop's local IP address
                String serverIP = "10.41.39.215";
                int serverPort = 5000;

                // Establish a socket connection
                Socket socket = new Socket(serverIP, serverPort);
                InputStream inputStream = socket.getInputStream();

                while (isRunning) {
                    // Read data from the server
                    byte[] buffer = new byte[1024];
                    int bytesRead = inputStream.read(buffer);
                    if (bytesRead == -1) break; // Connection closed

                    String serverTime = new String(buffer, 0, bytesRead);

                    // Update the UI with the received time
                    runOnUiThread(() -> textView.setText("Server Time: " + serverTime));
                }

                socket.close();
            } catch (Exception e) {
                e.printStackTrace();
                runOnUiThread(() -> textView.setText("Error: " + e.getMessage()));
            }
        }).start();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        // Stop the connection when the activity is destroyed
        isRunning = false;
    }
}

File: app/src/main/java/ro/pub/cs/systems/eim/practicaltest02v2/view/PracticalTest02v2ServerActivity.kt
package ro.pub.cs.systems.eim.practicaltest02v2.view

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import android.widget.TextView
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import ro.pub.cs.systems.eim.practicaltest02v2.R
import java.net.Socket

class PracticalTest02v2ServerActivity : AppCompatActivity() {

    private var isRunning = true

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_practical_test02v2_server)

        val textView = findViewById<TextView>(R.id.textView)

        GlobalScope.launch(Dispatchers.IO) {
            try {
                // Replace with your laptop's local IP address
                val serverIP = "10.41.39.215"
                val serverPort = 5000

                // Establish a socket connection
                val socket = Socket(serverIP, serverPort)
                val inputStream = socket.inputStream

                while (isActive && isRunning) {
                    // Read data from the server
                    val buffer = ByteArray(1024)
                    val bytesRead = inputStream.read(buffer)
                    if (bytesRead == -1) break // Connection closed

                    val serverTime = String(buffer, 0, bytesRead)

                    // Update the UI with the received time
                    withContext(Dispatchers.Main) {
                        textView.text = "Server Time: $serverTime"
                    }
                }

                socket.close()
            } catch (e: Exception) {
                e.printStackTrace()
                withContext(Dispatchers.Main) {
                    textView.text = "Error: ${e.message}"
                }
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        // Stop the connection when the activity is destroyed
        isRunning = false
    }
}
